File: MagicHax/android/app/src/main/java/me/nathanp/magiccreatures/model/LiveDatabase.java
package me.nathanp.magiccreatures.model;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.ValueEventListener;

public class LiveDatabase<T> extends LiveData<T> {
    private static final String TAG = "LiveDatabase";

    private DatabaseReference ref;
    private Class<T> clazz;
    private final LiveDatabaseListener listener = new LiveDatabaseListener();

    public LiveDatabase(DatabaseReference ref, Class<T> clazz) {
        this.ref = ref;
        this.clazz = clazz;
    }

    @Override
    protected void onActive() {
        ref.addValueEventListener(listener);
    }

    @Override
    protected void onInactive() {
        ref.removeEventListener(listener);
    }

    private class LiveDatabaseListener implements ValueEventListener {

        @Override
        public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
            setValue(dataSnapshot.getValue(clazz));
        }

        @Override
        public void onCancelled(@NonNull DatabaseError databaseError) {
            Log.e(TAG, databaseError.getMessage());
        }
    }
}

File: MagicHax/android/app/src/main/java/me/nathanp/magiccreatures/model/LiveDatabase.kt
package me.nathanp.magiccreatures.model

import android.util.Log
import androidx.lifecycle.LiveData
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.ValueEventListener

class LiveDatabase<T>(private val ref: DatabaseReference, private val clazz: Class<T>) : LiveData<T>() {
    private val TAG = "LiveDatabase"
    private val listener = LiveDatabaseListener()
    override fun onActive() {
        ref.addValueEventListener(listener)
    }

    override fun onInactive() {
        ref.removeEventListener(listener)
    }

    private inner class LiveDatabaseListener : ValueEventListener {
        override fun onDataChange(dataSnapshot: DataSnapshot) {
            value = dataSnapshot.getValue(clazz)
        }

        override fun onCancelled(databaseError: DatabaseError) {
            Log.e(TAG, databaseError.message)
        }
    }
}